multisig-wallet/project/multisig-contract/src/main.sw

```diff
@@ -1,212 +1,141 @@
-    /// spending
+    /// Used to add entropy into hashing of transaction to decrease the probability of collisions / double
+    /// spending.
     nonce: u64 = 0,
-    /// The number of approvals required in order to execture a Tx
+    /// The number of approvals required in order to execute a transaction.
     threshold: u64 = 0,
-    /// Number of approvals per user
-    weighting: StorageMap<Address, u64> = StorageMap {},
+    /// Number of approvals per user.
+    weighting: StorageMap<b256, u64> = StorageMap {},
 }
 
 impl MultiSignatureWallet for Contract {
-    /// The constructor initializes the necessary values and unlocks further functionlity
-    ///
-    /// # Panics
-    ///
-    /// - When the constructor is called more than once
-    /// - When the user address is the 0th address (0x00000...)
-    /// - When the threshold is set to 0
-    /// - When an owner has an approval weight of 0
     #[storage(read, write)]
-    fn constructor(users: [User; 25], threshold: u64) {
+    fn constructor(threshold: u64, users: Vec<User>) {
         require(storage.nonce == 0, InitError::CannotReinitialize);
         require(threshold != 0, InitError::ThresholdCannotBeZero);
 
         let mut user_index = 0;
-        while user_index < 25 {
-            require(Address::from(ZERO_B256) != users[user_index].identity, InitError::AddressCannotBeZero);
-            require(users[user_index].weight != 0, InitError::WeightingCannotBeZero);
-            storage.weighting.insert(users[user_index].identity, users[user_index].weight);
-            user_index = user_index + 1;
+        while user_index < users.len() {
+            require(ZERO_B256 != users.get(user_index).unwrap().address, InitError::AddressCannotBeZero);
+            require(users.get(user_index).unwrap().weight != 0, InitError::WeightingCannotBeZero);
+
+            storage.weighting.insert(users.get(user_index).unwrap().address, users.get(user_index).unwrap().weight);
+            user_index += 1;
         }
 
         storage.nonce = 1;
         storage.threshold = threshold;
     }
 
-    /// Executes a Tx formed from the `to, `value` and `data` parameters if the signatures meet the
-    /// threshold requirement
-    ///
-    /// # Panics
-    ///
-    /// - When the constructor has not been called to initialize the contract
-    /// - When the public key cannot be recovered from a signature
-    /// - When the recovered addresses are not in ascending order (0x1 < 0x2 < 0x3...)
-    /// - When the total approval count is less than the required threshold for execution
     #[storage(read, write)]
-    fn execute_transaction(to: Identity, value: u64, data: b256, signatures: [B512; 25]) {
+    fn execute_transaction(
+        data: b256,
+        signatures: Vec<SignatureInfo>,
+        to: Identity,
+        value: u64,
+    ) {
         require(storage.nonce != 0, InitError::NotInitialized);
```

Comment:
Should we also check that transactions are being executed in nonce order? There are scenarios where a signer would want to approve a transaction on the condition that all previous transactions in the queue have already been executed. Many multisig wallets (e.g. Gnosis Safe) have this requirement.

Reply:
As it stands, I don't think it's possible for transactions to be executed in anything but nonce order; the user never inputs a nonce value to either `transfer` or `execute_transaction`, the value is only ever called from storage when executing a transaction. 
A function to cancel an order ( by incrementing the nonce ) would be helpful; I'll add that in

Reply:
That makes sense. It's a feature that's rarely used in practice, but in a scenario where the signers don't all trust each other, being able to explicitly render a previously shared signature useless (by using its nonce) is necessary.